plugins/modem-manager/fu-mbim-qdu-updater.c

```diff
@@ -0,0 +1,482 @@
+/*
+ * Copyright (C) 2021 Jarvis Jiang <jarvis.w.jiang@gmail.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <string.h>
+#include <sys/stat.h>
+#include <stdio.h>
+
+#include "fu-common.h"
+#include "fu-mbim-qdu-updater.h"
+#include "fu-chunk.h"
+#include "fu-mm-utils.h"
+
+#if MBIM_CHECK_VERSION(1,25,3)
+#define FU_MBIM_QDU_MAX_OPEN_ATTEMPTS 8
+
+struct _FuMbimQduUpdater {
+	GObject		 parent_instance;
+	gchar		*mbim_port;
+	MbimDevice	*mbim_device;
+};
+
+G_DEFINE_TYPE (FuMbimQduUpdater, fu_mbim_qdu_updater, G_TYPE_OBJECT)
+
+typedef struct {
+	GMainLoop	*mainloop;
+	MbimDevice	*mbim_device;
+	GError		*error;
+	guint		 open_attempts;
+} OpenContext;
+
+static void fu_mbim_qdu_updater_mbim_device_open_attempt (OpenContext *ctx);
+
+static void
+fu_mbim_qdu_updater_mbim_device_open_ready (GObject *mbim_device, GAsyncResult *res, gpointer user_data)
+{
+	OpenContext *ctx = (OpenContext *) user_data;
+
+	if (!mbim_device_open_full_finish (MBIM_DEVICE (mbim_device), res, &ctx->error)) {
+		ctx->open_attempts--;
+		if (ctx->open_attempts == 0) {
+			g_clear_object (&ctx->mbim_device);
+			g_main_loop_quit (ctx->mainloop);
+			return;
+		}
+
+		/* retry */
+		g_clear_error (&ctx->error);
```

Comment:
maybe g_debug the failure?

Reply:
OK, I will g_debug the error, thanks for your suggestion.

Reply:
The `g_debug()` should not have a trailing `\n`